pandas/io/formats/info.py

```diff
@@ -209,147 +275,353 @@ def info(self) -> None:
+        return get_option("display.max_info_rows", len(self.data) + 1)
+
+    @property
+    def exceeds_info_cols(self) -> bool:
+        """Check if number of columns to be summarized does not exceed maximum."""
+        return bool(self.col_count > self.max_cols)
+
+    @property
+    def exceeds_info_rows(self) -> bool:
+        """Check if number of rows to be summarized does not exceed maximum."""
+        return bool(len(self.data) > self.max_rows)
+
+    @property
+    def col_count(self) -> int:
+        """Number of columns to be summarized."""
+        return self.info.col_count
+
+    def _initialize_max_cols(self, max_cols: Optional[int]) -> int:
         if max_cols is None:
-            max_cols = get_option("display.max_info_columns", col_count + 1)
+            return get_option("display.max_info_columns", self.col_count + 1)
+        return max_cols
 
-        max_rows = get_option("display.max_info_rows", len(self.data) + 1)
-
-        if self.null_counts is None:
-            show_counts = (col_count <= max_cols) and (len(self.data) < max_rows)
+    def _initialize_show_counts(self, show_counts: Optional[bool]) -> bool:
+        if show_counts is None:
+            return bool(not self.exceeds_info_cols and not self.exceeds_info_rows)
         else:
-            show_counts = self.null_counts
-        exceeds_info_cols = col_count > max_cols
+            return show_counts
+
+    def to_buffer(self, buf: Optional[IO[str]] = None) -> None:
+        """Save dataframe info into buffer."""
+        klass = self._select_table_builder()
+        table_builder = klass(info=self.info)
+        lines = table_builder.get_lines()
+        if buf is None:  # pragma: no cover
+            buf = sys.stdout
+        fmt.buffer_put_lines(buf, lines)
 
+    def _select_table_builder(self) -> Type["DataFrameTableBuilder"]:
+        """Select table builder based on verbosity and display settings."""
         if self.verbose:
-            self._verbose_repr(lines, ids, dtypes, show_counts)
+            return self._select_verbose_table_builder()
         elif self.verbose is False:  # specifically set to False, not necessarily None
-            self._non_verbose_repr(lines, ids)
+            return DataFrameTableBuilderNonVerbose
         else:
-            if exceeds_info_cols:
-                self._non_verbose_repr(lines, ids)
+            if self.exceeds_info_cols:
+                return DataFrameTableBuilderNonVerbose
             else:
-                self._verbose_repr(lines, ids, dtypes, show_counts)
+                return self._select_verbose_table_builder()
 
-        # groupby dtype.name to collect e.g. Categorical columns
-        counts = dtypes.value_counts().groupby(lambda x: x.name).sum()
-        collected_dtypes = [f"{k[0]}({k[1]:d})" for k in sorted(counts.items())]
-        lines.append(f"dtypes: {', '.join(collected_dtypes)}")
+    def _select_verbose_table_builder(self) -> Type["DataFrameTableBuilderVerbose"]:
```

Comment:
can you fold this into _select_table_builder, there is too much indirection here.

Reply:
Did that.